File: Data Structures & Algorithms/search-2d-matrix/submission-5.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        n,m = len(matrix), len(matrix[0])

        l = 0
        r = n * m - 1
        while l<=r:
            mid = l + (r-l)//2
            row, col = mid // m, mid % m
            if target < matrix[row][col]:
                r = mid -1
            elif target > matrix[row][col]:
                l = mid +1
            else: return True
    

        return False

File: Data Structures & Algorithms/n-queens/submission-5.java
class Solution {

    int col, diag, diag2;
 
    List<List<String>> res;
    public List<List<String>> solveNQueens(int n) {

        int[] tmp = new int[n];
        res = new ArrayList<>();
        backtrack(0, n, tmp);
        return res;

    }

    void backtrack(int row, int n, int[] tmp){
        if(row==n) {
            List<String> board = new ArrayList<>();
            for(int i=0;i<n;i++){
                char[] line = new char[n];
                Arrays.fill(line, '.');
                line[tmp[i]] = 'Q';

                board.add(new String(line));
            }
            res.add(board);
            return;
        }

        for(int i=0;i<n;i++){
            if((col&(1<<i))>0 ||(diag&(1<<(row+i)))>0 || (diag2&(1<<(row-i+n)))>0) continue;
            col^=(1<<i);
            diag^=(1<<(row+i));
            diag2^=(1<<(row-i+n));
            tmp[row]=i;
            backtrack(row+1, n, tmp);

            col^=(1<<i);
            diag^=(1<<(row+i));
            diag2^=(1<<(row-i+n));
        }
    }
}
